Browser: Convert file access to Core::Stream

AtkinsSJ opened the pull request:
Thought I should help out a bit with converting things, so here is a patch with all of Browser's Core::File usage replaced with Core::Stream.

cc: @sin-ack

sin-ack commented on Userland/Applications/Browser/BrowserWindow.cpp:
```diff
@@ -262,31 +263,40 @@ void BrowserWindow::build_menus()
+    if (!maybe_search_engines_file.is_error()) {
+        auto& search_engines_file = maybe_search_engines_file.value();
+        auto file_size = search_engines_file->size().release_value_but_fixme_should_propagate_errors();
+        auto maybe_buffer = ByteBuffer::create_uninitialized(file_size);
+        if (maybe_buffer.has_value()) {
```
In these cases I just don't bother wrapping in an if and just `.release_value()` instead, with the hope that eventually we'll get around to having it be an ErrorOr API and fix the billion usages. :^)

AtkinsSJ replied:
I did that initially but didn't like that release_value() hides that there should be an error check - someone who comes along later might not notice, whereas this makes it clear. I agree on ByteBuffer though, maybe I should get on and make that return errors since it was bothering me.

sin-ack commented on Userland/Applications/Browser/BrowserWindow.cpp:
```diff
@@ -262,31 +263,40 @@ void BrowserWindow::build_menus()
+            auto buffer = maybe_buffer.release_value();
+            if (search_engines_file->read_or_error(buffer)) {
+                StringView buffer_contents { buffer.bytes() };
+                if (auto maybe_json = JsonValue::from_string(buffer_contents); !maybe_json.is_error() && maybe_json.value().is_array()) {
```
Hmmm, maybe a little much to be cramped in an if statement condition?

sin-ack commented on Userland/Applications/Browser/main.cpp:
```diff
@@ -70,14 +70,23 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+            auto ad_filter_list = maybe_ad_filter_list.release_value();
+            auto buffer = maybe_buffer.release_value();
+
+            auto can_read_line = ad_filter_list->can_read_line();
+            while (!can_read_line.is_error() && can_read_line.value()) {
```
Since this is in `serenity_main`, I suggest moving this out to a new function so that you can enjoy the benefits of TRY while not exiting if anything fails. You could do something like:

```c++
if (auto result = read_content_filters(); result.is_error()) {
    dbgln("Oh no, I couldn't read the content filters: {}", result.error());
}
```

AtkinsSJ replied:
Good idea, I did something similar for search-engines too.

sin-ack commented on Userland/Libraries/LibProtocol/Request.cpp:
```diff
@@ -54,6 +54,40 @@ void Request::stream_into(OutputStream& stream)
     };
 }
 
+void Request::stream_into(Core::Stream::Stream& stream)
```
It's unfortunate that `AK::OutputStream` is incompatible with `Core::Stream` classes, but its API is a little baroque and not TRY-friendly. Perhaps in the future we can rectify that.

sin-ack reviewed (changes requested):
Looks mostly good, a few nits.

davidot commented on Userland/Applications/Browser/BrowserWindow.cpp:
```diff
@@ -262,31 +263,40 @@ void BrowserWindow::build_menus()
-                if (g_search_engine == url_format) {
-                    action->set_checked(true);
-                    search_engine_set = true;
+    auto maybe_search_engines_file = Core::Stream::File::open(Browser::search_engines_file_path(), Core::Stream::OpenMode::Read);
```
This gets indented quite far to the right and any error state is not handled explicitly.
Perhaps it would be more readable if you put it in a lambda which you immediately call.
Something like:
```cpp

[&] {
    auto code_here;
    if (maybe_search_engines_file.is_error())
        return;
     
     if (some_other_error())
        return;
    
     auto more_code;

    auto json = maybe_json.release_value().as_array();
    for (auto& json_item : json.values()) {
        if (!json_item.is_object())
            continue;
        auto search_engine = json_item.as_object();
        // etc.
    }
}();
```

Although I do realize that it may look like you return from the outer function...

AtkinsSJ replied:
Thanks. Yeah I've actually decided to wrap the search-engines code into a method that returns ErrorOr so I can TRY inside it, it's a lot better than what I had.

davidot replied:
Of course, that's even better.

AtkinsSJ commented:
Thanks to you both! :^)

Since last push:
- Extracted the search-engine and content-filter loading code into separate functions that return `ErrorOr<void>`, so they can use `TRY()` for error handling.
- Corrected creation of content-filter StringViews. I accidentally was creating StringViews for the whole buffer, not just the part that was read into, so it was returning garbage values.

sin-ack reviewed (approved):
LGTM

AtkinsSJ commented:
Fixed a merge conflict.
